### PR TITLE
feat: add projection to top level cypher directive

### PR DESCRIPTION
**Pull request by danstarns**
# Description

Missing functionality from `neo4j-graphql-js`. Resolves https://github.com/neo4j/graphql/issues/227. 

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [ ] TCK tests have been updated
- [x] Integration tests have been updated
- [x] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by mathix420** [comment]
Great job! Hope this gets approved soon!

**Comment by mathix420** [comment]
Just saw that the issue is still occuring in case you are using a custom resolver.

**Exemple:**
```graphql
type Video {
    id: ID! @id
    title: String!
    likes: Int! @cypher(statement:"MATCH (n)-[:LIKES]->(this) RETURN count(n)")
    ...
}

type Mutation {
    feedbackVideo(id: ID!, vote: Feedback!): Video! @requireAuth
}
```
```typescript
async feedbackVideo(_parent: any, { id, vote }: { id: string, vote: string }, { driver, auth }: Context): Promise<any> {
    const session = driver.session();
    let query: string;

    switch (vote) {
      case 'like':
        query = `
          MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
          OPTIONAL MATCH (u)-[r:DISLIKES]->(v)
          DELETE r
          MERGE (u)-[:LIKES]->(v)
          RETURN v
        `
        break;

      case 'dislike':
        query = `
          MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
          OPTIONAL MATCH (u)-[r:LIKES]->(v)
          DELETE r
          MERGE (u)-[:DISLIKES]->(v)
          RETURN v
        `
        break;

      default:
        query = `
          MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
          OPTIONAL MATCH (u)-[r:LIKES | DISLIKES]->(v)
          DELETE r
          RETURN v
        `
        break
    }

    const res = await session.run(query);
    const [ video ] = res.records;

    return video.toObject().v.properties;
  },
```

Then, this mutation
```graphql
mutation like {
  feedbackVideo(id: "40ac5ae5-193f-45b3-94c9-603875ea10db", vote: like) {
    id
    title
    likes
  }
}
```
Return this error
```
Cannot return null for non-nullable field Video.likes.
```

**Comment by jaedag** [comment]
Hi 
Is there any sort of timeline for when this PR will be merged?

**Comment by danstarns** [comment]
> Just saw that the issue is still occuring in case you are using a custom resolver.
> 
> **Exemple:**
> 
> ```graphql
> type Video {
>     id: ID! @id
>     title: String!
>     likes: Int! @cypher(statement:"MATCH (n)-[:LIKES]->(this) RETURN count(n)")
>     ...
> }
> 
> type Mutation {
>     feedbackVideo(id: ID!, vote: Feedback!): Video! @requireAuth
> }
> ```
> 
> ```ts
> async feedbackVideo(_parent: any, { id, vote }: { id: string, vote: string }, { driver, auth }: Context): Promise<any> {
>     const session = driver.session();
>     let query: string;
> 
>     switch (vote) {
>       case 'like':
>         query = `
>           MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
>           OPTIONAL MATCH (u)-[r:DISLIKES]->(v)
>           DELETE r
>           MERGE (u)-[:LIKES]->(v)
>           RETURN v
>         `
>         break;
> 
>       case 'dislike':
>         query = `
>           MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
>           OPTIONAL MATCH (u)-[r:LIKES]->(v)
>           DELETE r
>           MERGE (u)-[:DISLIKES]->(v)
>           RETURN v
>         `
>         break;
> 
>       default:
>         query = `
>           MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
>           OPTIONAL MATCH (u)-[r:LIKES | DISLIKES]->(v)
>           DELETE r
>           RETURN v
>         `
>         break
>     }
> 
>     const res = await session.run(query);
>     const [ video ] = res.records;
> 
>     return video.toObject().v.properties;
>   },
> ```
> 
> Then, this mutation
> 
> ```graphql
> mutation like {
>   feedbackVideo(id: "40ac5ae5-193f-45b3-94c9-603875ea10db", vote: like) {
>     id
>     title
>     likes
>   }
> }
> ```
> 
> Return this error
> 
> ```
> Cannot return null for non-nullable field Video.likes.
> ```

We can reconcile your custom javascript code with the single cypher statement. If you are using a custom Javascript function you should handle the tree walking yourself.

**Comment by mathix420** [comment]
> > Just saw that the issue is still occuring in case you are using a custom resolver.
> > **Exemple:**
> > ```graphql
> > type Video {
> >     id: ID! @id
> >     title: String!
> >     likes: Int! @cypher(statement:"MATCH (n)-[:LIKES]->(this) RETURN count(n)")
> >     ...
> > }
> > 
> > type Mutation {
> >     feedbackVideo(id: ID!, vote: Feedback!): Video! @requireAuth
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```ts
> > async feedbackVideo(_parent: any, { id, vote }: { id: string, vote: string }, { driver, auth }: Context): Promise<any> {
> >     const session = driver.session();
> >     let query: string;
> > 
> >     switch (vote) {
> >       case 'like':
> >         query = `
> >           MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
> >           OPTIONAL MATCH (u)-[r:DISLIKES]->(v)
> >           DELETE r
> >           MERGE (u)-[:LIKES]->(v)
> >           RETURN v
> >         `
> >         break;
> > 
> >       case 'dislike':
> >         query = `
> >           MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
> >           OPTIONAL MATCH (u)-[r:LIKES]->(v)
> >           DELETE r
> >           MERGE (u)-[:DISLIKES]->(v)
> >           RETURN v
> >         `
> >         break;
> > 
> >       default:
> >         query = `
> >           MATCH (u:User {id: '${auth.jwt.sub}'}), (v:Video {id: '${videoId}'})
> >           OPTIONAL MATCH (u)-[r:LIKES | DISLIKES]->(v)
> >           DELETE r
> >           RETURN v
> >         `
> >         break
> >     }
> > 
> >     const res = await session.run(query);
> >     const [ video ] = res.records;
> > 
> >     return video.toObject().v.properties;
> >   },
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then, this mutation
> > ```graphql
> > mutation like {
> >   feedbackVideo(id: "40ac5ae5-193f-45b3-94c9-603875ea10db", vote: like) {
> >     id
> >     title
> >     likes
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Return this error
> > ```
> > Cannot return null for non-nullable field Video.likes.
> > ```
> 
> We can reconcile your custom javascript code with the single cypher statement. If you are using a custom Javascript function you should handle the tree walking yourself.

I see, I already did this as a quick fix.
But I'm wondering now wouldn't be a good idea to allow custom resolver to return cypher queries as text for them to be interpreted as the `@cypher` directive does. Or maybe just a wrapper?

Not sure if I'm clear in my explanations :sweat_smile:

**Comment by danstarns** [comment]
@mathix420 Check out this workaround, using the OGM, here https://github.com/neo4j/graphql/issues/227#issuecomment-852819253 it could solve your issue.

**Comment by mathix420** [comment]
@danstarns thanks I will have a look!

File: packages/graphql/tests/integration/custom-resolvers-int.test.ts [hunk 1]
```diff
@@ -236,7 +235,7 @@ describe("Custom Resolvers", () => {
                         typeDefs = `
                             type Query {
                                 test: ${type}! @cypher(statement: """
-                                RETURN \"${id}\"
+                                RETURN \\"${id}\\"
```
Comment (darrellwarde) [inline packages/graphql/tests/integration/custom-resolvers-int.test.ts:238]: Out of interest, why is the extra escaping necessary here? Are we passing this complexity onto users?

**Review by darrellwarde — CHANGES_REQUESTED** [review]
This looks good, and I'm sure it works well based on the good number of integration tests.

Personally, I would really like to see the Cypher TCK tests for the `@cypher` directive to be updated to use this new feature, or add a new test for it. This will help me and other developers on the library to easily see what is output from these changes.

File: packages/graphql/tests/integration/custom-resolvers-int.test.ts (hunk 1 above)
Comment (danstarns) [reply packages/graphql/tests/integration/custom-resolvers-int.test.ts:238]: Now we are putting the given cypher query inside either apoc.runFirstColumn or apoc.doIt and so, with your highlighted, the first braces are to escape GraphQL and the second is to escape inside Cypher. 

```
CALL apoc.doIt("RETURN \"some-id\"")
```

**Comment by danstarns** [comment]
> This looks good, and I'm sure it works well based on the good number of integration tests.
> 
> Personally, I would really like to see the Cypher TCK tests for the `@cypher` directive to be updated to use this new feature, or add a new test for it. This will help me and other developers on the library to easily see what is output from these changes.

I understand your needs for TCK here however, this PR includes integration tests and in my opinion, they are satisfactory. It was a conscious decision not to include TCK due to the state of the TCK tests, things get rather messy and complicated adding the functionally to test specifically 'top-level' cypher directives.

**Comment by darrellwarde** [comment]
> > This looks good, and I'm sure it works well based on the good number of integration tests.
> > Personally, I would really like to see the Cypher TCK tests for the `@cypher` directive to be updated to use this new feature, or add a new test for it. This will help me and other developers on the library to easily see what is output from these changes.
> 
> I understand your needs for TCK here however, this PR includes integration tests and in my opinion, they are satisfactory. It was a conscious decision not to include TCK due to the state of the TCK tests, things get rather messy and complicated adding the functionally to test specifically 'top-level' cypher directives.

I agree that the integration tests cover the functionality side of things very well, but I'm not asking for TCK tests to validate functionality, but as a developer aid to help visualise what the effect of these changes are on the output Cypher. I get that the TCK tests definitely need a bit of a spring clean, but we can't neglect them in the meantime.

> things get rather messy and complicated adding the functionally to test specifically 'top-level' cypher directives.

However, unsure what you mean by this - are you saying that there would be additional changes needed to the actual TCK framework in order to support writing these tests?

**Review by darrellwarde — APPROVED** [review]
Based on our discussion regarding the complexity of adding TCK tests here, happy to approve. 🙂

**Pull request merged**